daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala

```diff
@@ -0,0 +1,651 @@
+    case ne: Node.NodeExercises[_, _] =>
+      tweakOptKeyMaintainers.run(ne.key).map { x => ne.copy(key = x) }
+    case _ => Nil
+  }
+  private def tweakExerciseByKey(whenVersion: TransactionVersion => Boolean) = Tweak.single[Node] {
+    case ne: Node.NodeExercises[_, _] =>
+      if (whenVersion(ne.version)) Some(ne.copy(byKey = changeBoolean(ne.byKey)))
+      else None
+    case _ => None
+  }
+  private val tweakExerciseVersion = Tweak.single[Node] {
+    case ne: Node.NodeExercises[_, _] => Some(ne.copy(version = changeVersion(ne.version)))
+    case _ => None
+  }
+
+  private val sigExeTweaks =
+    Map(
+      "tweakExerciseTargetCoid" -> tweakExerciseTargetCoid,
+      "tweakExerciseTemplateId" -> tweakExerciseTemplateId,
+      "tweakExerciseChoiceId" -> tweakExerciseChoiceId,
+      "tweakExerciseConsuming" -> tweakExerciseConsuming,
+      "tweakExerciseActingParties" -> tweakExerciseActingParties,
+      "tweakExerciseChosenValue" -> tweakExerciseChosenValue,
+      "tweakExerciseStakeholders" -> tweakExerciseStakeholders,
+      "tweakExerciseSignatories" -> tweakExerciseSignatories,
+      "tweakExerciseChoiceObservers" -> tweakExerciseChoiceObservers,
+      "tweakExerciseExerciseResult(Some)" -> tweakExerciseExerciseResultSome,
+      "tweakExerciseKey(Some)" -> tweakExerciseKey(tweakOptKeyMaintainersSome),
+      "tweakExerciseByKey(New Version)" -> tweakExerciseByKey(versionSinceMinByKey),
+      "tweakExerciseVersion" -> tweakExerciseVersion,
+    )
+
+  private val insigExeTweaks =
+    Map(
+      "tweakExerciseOptLocation" -> tweakExerciseOptLocation,
+      "tweakExerciseExerciseResult(None)" -> tweakExerciseExerciseResultNone,
+      "tweakExerciseKey(None)" -> tweakExerciseKey(tweakOptKeyMaintainersNone),
+      "tweakExerciseByKey(Old Version)" -> tweakExerciseByKey(versionBeforeMinByKey),
+    )
+
+  //--[significant and insignificant tx tweaks]--
+
+  private def tweakTxNodes(tweakNode: Tweak[Node]) = Tweak[VTX] { vtx =>
+    // tweak any node in a transaction
+    vtx.transaction match {
+      case GenTransaction(nodeMapA, roots) =>
+        for {
+          nid <- nodeMapA.keys.toList
+          nodeB <- tweakNode.run(nodeMapA(nid))
+        } yield {
+          val nodeMapB = nodeMapA + (nid -> nodeB)
+          val txB = GenTransaction(nodeMapB, roots)
+          TransactionVersion.asVersionedTransaction(txB)
+        }
+    }
+  }
+
+  private def significantTweaks: Map[String, Tweak[VTX]] = {
+    (sigCreateTweaks ++ sigFetchTweaks ++ sigLookupTweaks ++ sigExeTweaks)
+      .map { case (name, tw) => (name, tweakTxNodes(tw)) }
+  }
+
+  private def insignificantTweaks: Map[String, Tweak[VTX]] = {
+    (insigCreateTweaks ++ insigFetchTweaks ++ insigLookupTweaks ++ insigExeTweaks)
+      .map { case (name, tw) => (name, tweakTxNodes(tw)) }
+  }
+
+  //--[per tweak tests]--
+
+  "Significant tweaks" - {
+    var count = 0
+    significantTweaks.foreach { case (name, tweak) =>
```

Comment:
I'll have a look at this...

Reply:
`forEvery` is already doing a lot of the stuff for you so you don't have to do it yourself.  For example, a nested `forEvery` here would simply report the `txA` and `txB` in a nice way as part of the failure message, so you would not have to jam it into the `fail` string.

And _that_ makes it so that you can simply use `shouldBe` or `inside` without using `fail` at all, and still getting all the stuff you might need to know to understand failures reported.